### ISSUE TITLE
Support expand

### ISSUE THREAD

**Issue by csarofeen** [csarofeen/pytorch#1722]
### 🚀 The feature, motivation and pitch

An expand operation in PyTorch can take a tensor:
`size = [3, 17, 1, 80]` and `stride = [1360, 80, 80, 1]`
and physically expand the size-1 dimension so that:
`size = [3, 17, 99, 80]` and `stride = [1360, 80, 0, 1]`

Today nvFuser cannot handle these types of situations and will not expand a dimension, as it only broadcasts dimensions in the way that the dimension has `size == 1` and it doesn't matter what the stride is.

nvFuser also doesn't have a concept of stride besides what's passed in from a user, for example it can't deduce its own stride ordering of tensors. nvFuser only has the concept of contiguity, where contiguity == false means nvFuser has to generate code like `T0.stride[1]` in order to use the runtime passed in strides, but has no concept of what that value is.

The proposal is to somehow not change the overall internal representation of broadcast in nvFuser, but to be able to associate an expanded extent with a particular iteration domain.

The two approaches I've thought of to do this is:
1. Directly bury this "expanded" size into the iteration domain as a new property that's only relevant if an iteration domain is broadcasted.
2. Simply have this information passed into the fusion to hold onto for when we go to execution as it only matters when allocating outputs.

On a related note I don't recall how we deal with allocation and indexing into output tensors. For input tensors we have strided broadcast which indicates there's a stride associated with the broadcast dimension, broadcast without stride assumes there is no stride index. This is only important when indexing into global memory.

**Comment by rdspring1** [comment]
Linking other related issues:
https://github.com/csarofeen/pytorch/issues/1626
https://github.com/csarofeen/pytorch/issues/1507

**Comment by csarofeen** [comment]
Cross posting some conversation on this:
If we have something like:
`y = gelu(e.expand(...))` we would need to track the concrete size.

If we have something like a tensor `T1[1]` expanded to 99 as `T1[99, stride=0]` it would be a hard error to do something like `... = T1[99, stride=0] + T2[100]`

**Comment by csarofeen** [comment]
We would definitely have to add some mechanisms to check consistency at runtime to make sure concretized domains like this are correct. Actually... we would probably already error during expression evaluation.

**Comment by rdspring1** [comment]
Could we make the expanded dimensions dynamic?

**Comment by csarofeen** [comment]
I don't see why not.

**Comment by csarofeen** [comment]
@naoyam one thing between the two approaches, is the extent will have to propagate as new operations are added. Maybe it does make sense then as an IterDomain property.

**Comment by naoyam** [comment]
> @naoyam one thing between the two approaches, is the extent will have to propagate as new operations are added. Maybe it does make sense then as an IterDomain property.

Agreed.

**Comment by zasdfgbnm** [comment]
Can `expand` be implemented as a composite op?
```python
t.expand(-1, -1, 5, 7, -1)
```
Should be equivalent to
```python
t + torch.zeros(1, 1, 5, 7, 1)
```
I think?

If we create a dummy binary op `take_first(x, y) -> x`, and add support for `zeros`, should existing infrastructure be sufficient to support expand?

**Comment by csarofeen** [comment]
It would support it, but it would immediately make the tensor dense which I'm trying to avoid. If someone expands a tensor, but we don't need to materialize that expansion until many ops further down, we may get into a situation where we don't even need to materialize the expand at all.

In other words if we have a tensor with dim `32` but stride `0`, and we never interact with another tensor that has a real physical dimension there, we wouldn't ever have to physically expand the tensor.

**Comment by jjsjann123** [comment]
I'm wondering how would this affect indexing, specifically dimension collapsing? Specifying a domain as broadcasted in the middle of a tensor breaks our assumption that all dimensions follow ascending stride order. So I imaging having a broadcasted dimension in the middle would break the dimension collapsing, unless we pre-permute input tensors.

**Comment by csarofeen** [comment]
I think we now have support to collapse through broadcast. CC @naoyam to check my memory.

**Comment by naoyam** [comment]
As long as a broadcast domain is set as contiguous, dimension collapsing, i.e., contig-merged indexing, should work. The main logic is in contiguity.cpp.

**Comment by csarofeen** [comment]
Oh, I thought we ignored the contiguity flag on broadcast dims, we should try that. Never mind I see the issue, as we'd have to signal contiguity across broadcast dims :-/

**Comment by jjsjann123** [comment]
closing with #1739 #1754
